synchronization/src/androidTest/java/de/cyface/synchronization/WifiSurveyorTest.java

```diff
@@ -0,0 +1,149 @@
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import android.accounts.Account;
+import android.accounts.AccountManager;
+import android.content.ContentResolver;
+import android.content.Context;
+import android.net.ConnectivityManager;
+import android.os.Bundle;
+
+import androidx.test.ext.junit.runners.AndroidJUnit4;
+import androidx.test.platform.app.InstrumentationRegistry;
+import de.cyface.testutils.SharedTestUtils;
+import de.cyface.utils.Validate;
+
+/**
+ * Tests the correct functionality of the {@link WiFiSurveyor} class.
+ * <p>
+ * The tests in this class require an emulator or a real device.
+ *
+ * @author Armin Schnabel
+ * @version 1.0.0
+ * @since 4.0.0
+ */
+@RunWith(AndroidJUnit4.class)
+public class WifiSurveyorTest {
+
+    /**
+     * An object of the class under test.
+     */
+    private WiFiSurveyor objectUnderTest;
+    /**
+     * The {@link AccountManager} to check which accounts are registered.
+     */
+    private AccountManager accountManager;
+
+    /**
+     * Initializes the properties for each test case individually.
+     */
+    @Before
+    public void setUp() {
+        Context context = InstrumentationRegistry.getInstrumentation().getTargetContext();
+        ConnectivityManager connectivityManager = (ConnectivityManager)context
+                .getSystemService(Context.CONNECTIVITY_SERVICE);
+        Validate.notNull(connectivityManager);
+
+        objectUnderTest = new WiFiSurveyor(context, connectivityManager, AUTHORITY, ACCOUNT_TYPE);
+
+        // Ensure reproducibility
+        accountManager = AccountManager.get(context);
+        SharedTestUtils.cleanupOldAccounts(accountManager, ACCOUNT_TYPE, AUTHORITY);
+    }
+
+    @After
+    public void tearDown() {
+        final Account[] oldAccounts = accountManager.getAccountsByType(ACCOUNT_TYPE);
+        if (oldAccounts.length > 0) {
+            for (Account oldAccount : oldAccounts) {
+                ContentResolver.removePeriodicSync(oldAccount, AUTHORITY, Bundle.EMPTY);
+                Validate.isTrue(accountManager.removeAccountExplicitly(oldAccount));
+            }
+        }
+        objectUnderTest = null;
+    }
+
+    /**
+     * Tests that marking the connection as syncable using the account flags works.
+     * <p>
+     * This test reproduced MOV-635 where the periodic sync flag did not change because syncAutomatically was not set.
+     * This bug was only reproducible in integration environment (device and emulator) but not as roboelectric test.
+     * <p>
+     * This test may be flaky on a read device when the network changes during the test.
+     */
+    @Test
+    public void testSetConnected() throws InterruptedException {
+
+        // Arrange
+        final Lock lock = new ReentrantLock();
+        final Condition condition = lock.newCondition();
```

Comment:
Wo wird denn die Condition wieder frei gegeben (`condition.signal()`)?

Reply:
Wieso muss ich die condition explizit freigeben?

> condition.await(1, TimeUnit.SECONDS); // 200ms was not enough for CI emulators

Ich habe das so interprediert, dass der Test somit wartet und erst nach der Sekunde prüft, ob die Flag geschrieben wurde. Um das zu beschleunigen könnte man natürlich alle paar Millisekunden prüfen ob das schon eher der Fall ist und die condition frühzeitig freigeben aber das wäre Overkill in dem Fall.

Viele andere Tests von uns geben auch kein Signal.
Ich dachte das await gibt nach der abgelaufenen Zeit ein Signal. Der Test läuft ja auch so wie ich mit das vorgestellt habe.

Reply:
Es sollte im Code eigentlich keine Stelle geben wo das Signal fehlt. Manchmal wird das innerhalb einer aufgerufenen Klasse aufgerufen. Man macht das üblicherweise in einem Callback.

Reply:
Danke für die Klarstellung. An anderen Stellen ist das auch der Fall.